document/encoding/msgpack/encoding_test.go

```diff
@@ -0,0 +1,300 @@
+package msgpack
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/genjidb/genji/document"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetByField(t *testing.T) {
+	doc := document.NewFieldBuffer().
+		Add("age", document.NewInt64Value(10)).
+		Add("address", document.NewNullValue()).
+		Add("name", document.NewTextValue("john")).
+		Add("d", document.NewDurationValue(10*time.Nanosecond)).
+		Add(strings.Repeat("a", 2<<5), document.NewBoolValue(true)).
+		Add(strings.Repeat("a", 2<<10), document.NewBoolValue(true)).
+		Add(strings.Repeat("a", 2<<20), document.NewBoolValue(true))
+
+	data, err := EncodeDocument(doc)
+	require.NoError(t, err)
+
+	d := EncodedDocument(data)
+
+	v, err := d.GetByField("age")
+	require.NoError(t, err)
+	require.Equal(t, document.NewInt64Value(10), v)
+
+	v, err = d.GetByField("address")
+	require.NoError(t, err)
+	require.Equal(t, document.NewNullValue(), v)
+
+	v, err = d.GetByField("name")
+	require.NoError(t, err)
+	require.Equal(t, document.NewTextValue("john"), v)
+
+	v, err = d.GetByField("doesnexists")
+	require.Equal(t, err, document.ErrFieldNotFound)
+
+	v, err = d.GetByField("d")
+	require.NoError(t, err)
+	require.Equal(t, document.NewDurationValue(10*time.Nanosecond), v)
+
+	v, err = d.GetByField(strings.Repeat("a", 2<<20))
+	require.NoError(t, err)
+	require.Equal(t, document.NewBoolValue(true), v)
+}
+
+func TestGetByIndex(t *testing.T) {
+	arr := document.NewValueBuffer().
+		Append(document.NewInt64Value(10)).
+		Append(document.NewNullValue()).
+		Append(document.NewTextValue("john")).
+		Append(document.NewDurationValue(10 * time.Nanosecond))
+
+	data, err := EncodeArray(arr)
+	require.NoError(t, err)
+
+	a := EncodedArray(data)
+
+	v, err := a.GetByIndex(0)
+	require.NoError(t, err)
+	require.Equal(t, document.NewInt64Value(10), v)
+
+	v, err = a.GetByIndex(1)
+	require.NoError(t, err)
+	require.Equal(t, document.NewNullValue(), v)
+
+	v, err = a.GetByIndex(2)
+	require.NoError(t, err)
+	require.Equal(t, document.NewTextValue("john"), v)
+
+	v, err = a.GetByIndex(1000)
+	require.Equal(t, err, document.ErrValueNotFound)
+}
+
+func TestEncodeDecode(t *testing.T) {
+	userMapDoc, err := document.NewFromMap(map[string]interface{}{
+		"age":  10,
+		"name": "john",
+	})
+	require.NoError(t, err)
+
+	addressMapDoc, err := document.NewFromMap(map[string]string{
+		"city":    "Ajaccio",
+		"country": "France",
+	})
+	require.NoError(t, err)
+
+	tests := []struct {
+		name     string
+		d        document.Document
+		expected string
+	}{
+		{
+			"document.FieldBuffer",
+			document.NewFieldBuffer().
+				Add("age", document.NewInt64Value(10)).
+				Add("name", document.NewTextValue("john")),
+			`{"age": 10, "name": "john"}`,
+		},
+		{
+			"Map",
+			userMapDoc,
```

Comment:
🌷 I think we can remove `userMapDoc` and use directly `document.NewFromMap` here instead.

Reply:
`document.NewFromMap` returns an error, right? That's why we are creating it first, then assigning it here.
Or do you have something else in mind?

Reply:
Oops, indeed, i missed that.